### DISCUSSION TITLE
A universal MacOS build of current code

### DISCUSSION THREAD

**Discussion by emendelson** [joncampbell123/dosbox-x#4465] [Show and tell]
For a project I'm working on, I wanted to have a universal (Intel+AppleSilicon) build of DOSBox-X, but was frustrated by the issues described here: https://github.com/joncampbell123/dosbox-x/issues/4438 and other problems. Using macdylibbundler (https://github.com/auriamg/macdylibbundler) I was able to put together a universal app, by combining a version built on Intel and a version built on an M2 Mac, and using `lipo` to merge the executables.

If anyone's interested, here it is (SDL2 only) in an app that's notarized. It should work on Intel Macs running Mojave or later (I built the Intel version under Mojave) and on any Apple Silicon Mac:

https://www.dropbox.com/scl/fi/27wv918rl9ckcjsjs03hl/DOSBox-X-Universal.zip?rlkey=qwgnftklgmjoizt9jznofce97&dl=0

Please let me know of any problems with this.

**Comment by Shrapnel21** [comment]
It works fine! Thank you! =)

**Comment by emendelson** [comment]
And here's an SDL1 version (not fully tested, but should be OK):

https://www.dropbox.com/scl/fi/d2rpidf5uo39kpww3uknl/DOSBox-X-SDL1-Universal.zip?rlkey=kzf8021qizexcj48nnj6t9nvr&dl=0

**Comment by Torinde** [comment]
Related: #4054

**Comment by emendelson** [comment]
I've been asked to explain how to create a universal macOS build of DOSBox-X. This is a brief answer, but it should be enough for anyone. 

1. Build for SDL1 on an Intel machine running the earliest version that you want to support. Make sure to run `make dosbox-x.app` after running the build script in order to create a DOSBox-X.app.
2. If you want, do the same for SDL2.
3. Build for SDL1 on an ARM Mac running the earliest version that you want to support. Again, make sure to run `make dosbox-x.app` after running the build script.
4. If you want, do the same for SDL2.
5. Copy the Intel builds to your ARM Mac (or the other way around; it doesn't matter).
6. Extract the `dosbox-x` executable from the Contents/MacOS folder in the Intel SDL1 build and copy it to the desktop, naming it something like `dosbox-Intel`.
7. Extract the `dosbox-x` executable from the ARM SDL1 build and copy it to the desktop, naming it something like `dosbox-ARM`.
8. If you don't want to look up the `lipo` syntax for yourself, download and use this Applescript that runs `lipo` for you:

[https://mendelson.org/LipoHelper.zip](https://mendelson.org/LipoHelper.zip)

9. Run that Applescript and create a universal executable. 
10. Copy that executable into one of the builds (it doesn't matter which), replacing the `dosbox-x` executable that's already there.
11. If you copied the universal executable into the Intel build, then copy the contents of the `arm64` folder from the ARM build into the empty arm64 folder of the Intel build. (Or the reverse: you want an app bundle that has the necessary files in **both** the arm64 and intel folders in the MacOS folder.)
12. The combined executable should now be universal.

I may have left out one or more important steps. If so, please comment here.
